RavenDB-12535: Asynchronous Global Prefetcher

redknightlois opened the pull request:
PrefetchVirtualMemory is a quite costly kernel call. We should just enqueue this on a dedicated worker thread and continue with our work, expecting the OS to do the right thing and when we need that memory we already have the memory in place.

ayende commented on src/Sparrow/Collections/RingBuffer.cs:
```diff
@@ -0,0 +1,136 @@
+                if (cidx - sidx >= this._size)
+                    return false; // No space to do anything
+
+                Interlocked.CompareExchange(ref _currentIdx, _currentIdx + 1, _currentIdx);
```
Shouldn't you be checking if this succeeded?

redknightlois replied:
This is a reservation, I have code here that is using `Interlocked.Increment` instead. Needed to run slow tests, that is why I did this PR...

ayende commented on src/Sparrow/Collections/RingBuffer.cs:
```diff
@@ -0,0 +1,136 @@
+                Interlocked.CompareExchange(ref _currentIdx, _currentIdx + 1, _currentIdx);
+                
+                // We check if we were successful 
+                if (Volatile.Read(ref _currentIdx) > cidx)
```
Isn't there a race here, why no check the return value from `CompareExchange` ?

redknightlois replied:
Not really. Because it works as a reservation system.

ayende commented on src/Voron/GlobalPrefetchingBehavior.cs:
```diff
@@ -0,0 +1,105 @@
+                {
+                    var commands = CommandQueue.Acquire();
+                    if (commands.IsEmpty)
+                        Thread.Sleep(1000); // If no command is in the queue, we are surely not busy. 
```
That is a bad idea. If we are sleeping and suddenly want to prefetch, that is going to _not_ prefetch and we have piecemeal I/O
Should use an event, instead.

redknightlois replied:
Nope. Because the caller is high frequency and we want to avoid the actual cost on the caller side. We dont care if we miss a few, in fact it goes way beyond that... we dont care if we miss a ton of them because we dont have enough space on the reservation queue.

ayende replied:
The caller is already protected by `ShouldPrefetch`, so I don't think we are going to see too many of that

redknightlois replied:
You will. Don't trust me, trust the profiling trace :)

redknightlois replied:
After some changes, the profiling trace suggests we can do wake up based on quantity. I tried though to get down the number of milliseconds down, with barely any change in hard page faults though.

ayende replied:
Can you show the numbers?

redknightlois replied:
180 hard faults per second either way.

redknightlois replied:
For reference we were in the 100ish range. But I didn't have time today to check the performance targets as I don't know the behavior yet of 4.2

redknightlois replied:
Implemented using a BlockingCollection instead. The wait was too much on average.

ayende commented on src/Voron/GlobalPrefetchingBehavior.cs:
```diff
@@ -0,0 +1,105 @@
+
+        private readonly Logger _log = LoggingSource.Instance.GetLogger<GlobalPrefetchingBehavior>("Global Prefetcher");
+         
+        public readonly SingleConsumerRingBuffer<PrefetchCommand> CommandQueue = new SingleConsumerRingBuffer<PrefetchCommand>(128);
```
Why do we want something like this?
It would be much easier to use a `ConcurrentQueue`, instead.
This is going to be triggering I/O, so I don't think we need to worry that much about efficiencies.

redknightlois replied:
Yes we do, because the one that is going to pay for it is the caller... The whole point of the optimization is to move the slow parts outside of the critical path, even at the expense of losing a few hits here and there.

redknightlois replied:
Using `BlockingCollection`

ayende commented on src/Sparrow/Platform/Win32/Win32MemoryMethods.cs:
```diff
@@ -1,9 +1,10 @@
 namespace Sparrow.Platform.Win32
 {
-    public static unsafe class Win32MemoryProtectMethods
+    public static unsafe class Win32MemoryMethods
```
why this change?

redknightlois replied:
Because it was a mess, all methods there are Memory methods, not protect methods. They happen to be useful for that, but we have many things scattered all over the place.

ayende replied:
Break it to a separate PR, it make it really hard to review the changes for this PR, which are already complex enough to require careful review

redknightlois replied:
Its just a rename, nothing more. Had to move stuff around to access that. So you will still see the changes in places that used that anyway. To me is a simple refactor rename.

ayende reviewed (changes requested):
This is very going to conflict with the librvnpal changes, pay attention to that.

redknightlois committed:
RavenDB-12535: Asynchronous Global Prefetcher

PrefetchVirtualMemory is a quite costly kernel call. We should just enqueue this on a dedicated worker thread and continue with our work, expecting the OS to do the right thing and when we need that memory we already have the memory in place.

redknightlois committed:
RavenDB-12535: Implement the Global Prefetcher with a blocking queue

In order to diminish as much as possible the time between when a page is needed and the time when the IO is initiated the Global Prefetcher thread will block waiting for new requests to come by.

ppekrol commented on src/Sparrow/Platform/Win32/Win32ProcessMethods.cs:
```diff
@@ -0,0 +1,15 @@
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Sparrow.Platform.Win32
```
Needs to be moved to Sparrow.Server

redknightlois replied:
Botched merge, that is not used. Removing

ppekrol commented on src/Sparrow/Collections/RingBuffer.cs:
```diff
@@ -0,0 +1,132 @@
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Sparrow.Collections
```
Better to move it to Sparrow.Server if this is not used in Raven.Client

redknightlois replied:
Botched merge, that is not used. removing.

redknightlois replied:
That is going to be used by RavenDB-12799 (https://issues.hibernatingrhinos.com/issue/RavenDB-12799) on `Sparrow`

ayende replied:
What is this going to be used _in sparrow_ ?

ayende replied:
Okay, this is used by the logging infrastructure, which we want to keep in sparrow

redknightlois committed:
RavenDB-12535: Removed from wrongly rebase merge.

redknightlois committed:
RavenDB-12799: Implementation using a ring buffer backend.

ayende commented on src/Sparrow.Server/Platform/Win32/Win32MemoryProtectMethods.cs:
```diff
@@ -24,8 +25,12 @@ public static unsafe class Win32MemoryProtectMethods
+        //public static extern bool PrefetchVirtualMemory(IntPtr hProcess, UIntPtr NumberOfEntries, MemorySegment* VirtualAddresses, ulong Flags);
+
+        //[DllImport("kernel32.dll", SetLastError = true)]
+        //public static extern bool DiscardVirtualMemory(void* lpAddress, UIntPtr size);
```
Better to delete, rather than comment the code

redknightlois replied:
Yeah. Was merging and wanted to be sure it was not needed and forgot to delete afterwards

ayende commented on src/Sparrow/Collections/RingBuffer.cs:
```diff
@@ -0,0 +1,157 @@
+            int ticker = Interlocked.Increment(ref _currentIdx) - 1;
+
+            // Assign the value
+            ref var cl = ref _buffer[ticker % this._size];
```
The compiler can't assume power of 2 here, no?
Can we turn this to a shift?

redknightlois replied:
We can if we control the size. Easy change if we want to do so.

ayende commented on src/Sparrow/Collections/RingBuffer.cs:
```diff
@@ -0,0 +1,157 @@
+            Volatile.Write(ref cl.IsReady, true);
+            return true;
+
+            // Very rarely we will go through and try again, the only condition is when we have a huge contention and traffic.
```
what make us go through here again?

redknightlois replied:
The ring buffer works in a acquire, release mode. You first acquire, do what you need to do and then release. On very small buffers with very high traffic you could end up in a situation where you have hit the bounded size. You end up there when that happens.

ayende commented on src/Voron/Impl/Paging/AbstractPager.cs:
```diff
@@ -529,45 +542,27 @@ public void MaybePrefetchMemory(long pageNumber, int pagesToPrefetch)
 
-            int prefetchIdx = 0;
-            var toPrefetch = stackalloc PalDefinitions.PrefetchRanges[StackSpace];
+            var command = default(PalDefinitions.PrefetchRanges);
```
```suggestion
            PalDefinitions.PrefetchRanges command = default;
```

ayende replied:
This looks clearer (minor)

ayende commented on src/Voron/Impl/Paging/RvnMemoryMapPager.cs:
```diff
@@ -150,11 +150,11 @@ protected override void DisposeInternal()
-            rvn_prefetch_ranges(list, count, out _);
-            // we explicitly ignore the return code here, this is optimization only
-        }
+        //protected internal override void PrefetchRanges(PrefetchRanges* list, int count)
```
Delete commented code

ayende commented on src/Voron/StorageEnvironment.cs:
```diff
@@ -158,6 +158,9 @@ public StorageEnvironment(StorageEnvironmentOptions options)
                     SuggestSyncDataFile(); // let's suggest syncing data file after the recovery
+
+                // Ensure we are always have the prefetcher available.
+                var prefetcher = GlobalPrefetchingBehavior.GlobalPrefetcher.Value;
```
```suggestion
                GC.KeepAlive(GlobalPrefetchingBehavior.GlobalPrefetcher.Value);
```

We use this commonly, and it is guaranteed that the compiler won't elide this

redknightlois replied:
Ok, for readability purposes it makes a lot of sense. But, there is no way thr compiler could elide it, because of the static side effects. If that would happen, there are a few invariants violated (major regression).

redknightlois committed:
RavenDB-12799: Improved codegen for the ring buffer using power of 2

redknightlois commented:
Done. Added power of 2 assumptions and therefore better codegen.

arekpalinski commented on test/FastTests/Sparrow/RingBuffer.cs:
```diff
@@ -0,0 +1,147 @@
+
+namespace FastTests.Sparrow
+{
+    public class RingBufferTests
```
Should inherit from `NoDisposalNeeded`

ppekrol committed:
RavenDB-12535 cleanup

ppekrol committed:
RavenDB-12535 added an option to disable prefetching mechanism

ppekrol commented:
Rebased, did a cleanup and added an option to disable prefetching

ppekrol commented:
Why it needs to be public? it is used only by us internally.

redknightlois commented:
That is in sparrow, I am working on POCs on voron that use that structure

ppekrol commented:
Internals are visible to Voron and Sparrow.Server

ppekrol committed:
RavenDB-12535 ring buffer tests do not need to be disposed

redknightlois commented:
Didn't know that we have setup a friend. Great

redknightlois committed:
RavenDB-12535: Addressed a few other reviews.
